game/src/debug/path_counter.rs

```diff
@@ -0,0 +1,141 @@
+// A state to count the number of trips that will cross different roads.
+
+use crate::app::{App, ShowEverything};
+use crate::common::{ColorLegend, ColorNetwork, CommonState};
+use crate::game::{State, Transition};
+use crate::helpers::ID;
+use abstutil::Counter;
+use map_model::{IntersectionID, PathStep, RoadID, Traversable};
+use sim::DontDrawAgents;
+use widgetry::{
+    Btn, Color, Drawable, EventCtx, GfxCtx, HorizontalAlignment, Key, Line, Outcome, Panel, Text,
+    VerticalAlignment, Widget,
+};
+
+pub struct PathCounter {
+    panel: Panel,
+    unzoomed: Drawable,
+    zoomed: Drawable,
+    cnt: Counter<RoadID>,
+    tooltip: Option<Text>,
+}
+
+impl PathCounter {
+    pub fn demand_across_intersection(
+        ctx: &mut EventCtx,
+        app: &App,
+        i: IntersectionID,
+    ) -> Box<dyn State> {
+        let map = &app.primary.map;
+        let sim = &app.primary.sim;
+        let mut cnt = Counter::new();
+        // Find all active agents whose path crosses this intersection
+        for agent in sim.active_agents() {
+            if let Some(path) = sim.get_path(agent) {
+                if path.get_steps().iter().any(|step| match step {
+                    PathStep::Turn(t) => t.parent == i,
+                    _ => false,
+                }) {
+                    // Count what lanes they'll cross
+                    for step in path.get_steps() {
+                        if let Traversable::Lane(l) = step.as_traversable() {
+                            cnt.inc(map.get_l(l).parent);
+                        }
+                    }
+                }
+            }
+        }
+
+        let mut colorer = ColorNetwork::new(app);
+        // Highlight the intersection
+        colorer
+            .unzoomed
+            .push(Color::CYAN, map.get_i(i).polygon.clone());
+        colorer
+            .zoomed
+            .push(Color::CYAN.alpha(0.5), map.get_i(i).polygon.clone());
+
+        colorer.ranked_roads(cnt.clone(), &app.cs.good_to_bad_red);
```

Comment:
It's nice that you have a lot of this existing code infrastructure to do something like this relatively concisely. 👏👏👏

Reply:
Thanks! Lots of the dataviz is still hard to interpret -- for this one, the shades of red sometimes feels less clear than discrete color buckets, and I'm not sure displaying by rank and not absolute number even makes sense. But at least the code is consolidated, so we can improve these in the future easily.